Fix: Improper KeyError when channel created on a BlockingConnection whose server was shutdown

gst opened the pull request:
More specifically:

    return self._channels[channel_number]

will raise KeyError in such case because the channel will have already been removed from the connection _channels dict.

Using this fix you'll more correctly get ConnectionClosed exception.

gst commented:
see https://github.com/pika/pika/issues/659 for more details

vitaly-krugl commented:
Hi @gst, I am unable to reproduce the KeyError exception against pika master using instructions from #659. Here is my session:

In [1]: import pika

In [2]: c = pika.BlockingConnection()

Now, stop RabbitMQ via `rabbitmqctl stop` before executing `ch = c.channel()`

In [3]: ch = c.channel()
---------------------------------------------------------------------------
ConnectionClosed                          Traceback (most recent call last)
<ipython-input-3-204314744cf0> in <module>()
----> 1 ch = c.channel()

/Users/vkruglikov/ossdev/pika/pika/adapters/blocking_connection.pyc in channel(self, channel_number)
    705 
    706             # Drive I/O until Channel.Open-ok
--> 707             channel._flush_output(opened_args.is_ready)
    708 
    709 

/Users/vkruglikov/ossdev/pika/pika/adapters/blocking_connection.pyc in _flush_output(self, *waiters)
   1181         self._connection._flush_output(
   1182             self._channel_closed_by_broker_result.is_ready,
-> 1183             *waiters)
   1184 
   1185         if self._channel_closed_by_broker_result:

/Users/vkruglikov/ossdev/pika/pika/adapters/blocking_connection.pyc in _flush_output(self, *waiters)
    420                                     result)
    421                     raise exceptions.ConnectionClosed(result.reason_code,
--> 422                                                       result.reason_text)
    423                 elif not self._user_initiated_close:
    424                     # NOTE: unfortunately, upon socket error, on_close_callback

ConnectionClosed: (320, "CONNECTION_FORCED - broker forced connection closure with reason 'shutdown'")

gst commented:
Hi @vitaly-krugl 

I just do : 

```
(env) ✘ ~/work/public/python/pika [master|…746]
greg@hal  $ python -c 'import time, pika; print(pika); c = pika.BlockingConnection(); time.sleep(5) ; c.channel()'
<module 'pika' from '/mnt/DATA/greg/work/public/python/pika/pika/__init__.py'>
Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "/mnt/DATA/greg/work/public/python/pika/pika/adapters/blocking_connection.py", line 698, in channel
    channel_number=channel_number)
  File "/mnt/DATA/greg/work/public/python/pika/pika/connection.py", line 733, in channel
    return self._channels[channel_number]
KeyError: 1
(env) ✘ ~/work/public/python/pika [master|…746]
greg@hal  $ 
```

during the time.sleep(5) I shutdown rabbitmq server (with `sudo service rabbitmq-server stop`).

I'm on master:

```
$ git log -1
commit d8a782d97579cd96ed67ccfb55f63ca8fdafa199
Merge: 208b8b5 3ef1a16
Author: Gavin M. Roy <gavinmroy@gmail.com>
Date:   Mon Dec 7 18:24:57 2015 -0500

    Merge pull request #672 from dave-shawley/call-open-error-cb
    
    Call on open error callback for probable auth/access failures
```

vitaly-krugl commented:
@gst, I am beginning to think that there is a deeper problem here that is perhaps tied to pika on python 3.x or on the different platform than I am running. By the way, what is your environment? Are you running on Unix or Linux and on which version?

Would you mind reproducing the failure with DEBUG logging turned on and posting pika's complete log here? Turn on DEBUG logging before creating the pika connection
```
import logging
LOG_FORMAT = ('%(levelname) -10s %(asctime)s %(name) -30s %(funcName) '
              '-35s %(lineno) -5d: %(message)s')
logging.basicConfig(level=logging.DEBUG, format=LOG_FORMAT)
```


I am running on Mac OS X Yosemite 10.10.5 with python 2.7.10, and here is what I get (note my modified command that stops rabbitmq inline):

```
$ python -c 'import pika; print(pika); c = pika.BlockingConnection(); import subprocess; subprocess.check_call(["rabbitmqctl", "stop"]); c.channel()'
<module 'pika' from '/Users/vkruglikov/ossdev/pika/pika/__init__.pyc'>
Stopping and halting node rabbit@localhost ...
Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "/Users/vkruglikov/ossdev/pika/pika/adapters/blocking_connection.py", line 707, in channel
    channel._flush_output(opened_args.is_ready)
  File "/Users/vkruglikov/ossdev/pika/pika/adapters/blocking_connection.py", line 1183, in _flush_output
    *waiters)
  File "/Users/vkruglikov/ossdev/pika/pika/adapters/blocking_connection.py", line 422, in _flush_output
    result.reason_text)
pika.exceptions.ConnectionClosed: (320, "CONNECTION_FORCED - broker forced connection closure with reason 'shutdown'")
```

gst commented:
I'm on ubuntu 15.04 and its his system Python3.4(.3).
with my python2.7 I also get the ConnectionClosed : 
```
$ python2.7 -c 'import pika; print(pika); c = pika.BlockingConnection(); import subprocess; subprocess.check_call(["sudo", "rabbitmqctl", "stop"]); c.channel()' 
<module 'pika' from 'pika/__init__.py'>
Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "pika/adapters/blocking_connection.py", line 339, in __init__
    self._process_io_for_connection_setup()
  File "pika/adapters/blocking_connection.py", line 374, in _process_io_for_connection_setup
    self._open_error_result.is_ready)
  File "pika/adapters/blocking_connection.py", line 397, in _flush_output
    raise exceptions.ConnectionClosed()
pika.exceptions.ConnectionClosed
```
now what's funny is that with a self-compiled python3.5 I also get ConnectionClosed:

```
$ /opt/python/3.5.0/bin/python3 -c 'import pika; print(pika); c = pika.BlockingConnection(); import subprocess; subprocess.check_call(["sudo", "rabbitmqctl", "stop"]); c.channel()'
<module 'pika' from '/mnt/DATA/greg/work/public/python/pika/pika/__init__.py'>
Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "/mnt/DATA/greg/work/public/python/pika/pika/adapters/blocking_connection.py", line 339, in __init__
    self._process_io_for_connection_setup()
  File "/mnt/DATA/greg/work/public/python/pika/pika/adapters/blocking_connection.py", line 374, in _process_io_for_connection_setup
    self._open_error_result.is_ready)
  File "/mnt/DATA/greg/work/public/python/pika/pika/adapters/blocking_connection.py", line 397, in _flush_output
    raise exceptions.ConnectionClosed()
pika.exceptions.ConnectionClosed
```

guessing this could effectively be a problem with this specific python 3.4(.3?) version..

vitaly-krugl commented:
@gst, I think it's related to timing and that there really is a race condition in pika's implementation. Posting the DEBUG-level log would help me get a handle on it. Many thanks.

gst commented:
Here it is : 

```
greg@hal  $ sudo service rabbitmq-server start
(env) ✔ ~/work/public/python/pika [master|…831]
greg@hal  $ python -c "import logging ; LOG_FORMAT = ('%(levelname) -10s %(asctime)s %(name) -30s %(funcName) -35s %(lineno) -5d: %(message)s') ; logging.basicConfig(level=logging.DEBUG, format=LOG_FORMAT); import pika; print(pika); c = pika.BlockingConnection(); import subprocess; subprocess.check_call(['sudo', 'rabbitmqctl', 'stop']); c.channel()"
<module 'pika' from '/mnt/DATA/greg/work/public/python/pika/pika/__init__.py'>
DEBUG      2015-12-20 15:53:19,331 pika.adapters.select_connection _get_poller                          137 : Using EPollPoller
DEBUG      2015-12-20 15:53:19,331 pika.callback                  add                                  164 : Added: {'one_shot': False, 'only': None, 'callback': <bound method _CallbackResult.set_value_once of <pika.adapters.blocking_connection._CallbackResult object at 0x7f6ffb834cc8>>, 'arguments': None}
DEBUG      2015-12-20 15:53:19,331 pika.callback                  add                                  164 : Added: {'one_shot': False, 'only': None, 'callback': <bound method _CallbackResult.set_value_once of <pika.adapters.blocking_connection._CallbackResult object at 0x7f6ffb834ec8>>, 'arguments': None}
DEBUG      2015-12-20 15:53:19,331 pika.callback                  add                                  164 : Added: {'one_shot': False, 'only': None, 'callback': <bound method _CallbackResult.set_value_once of <pika.adapters.blocking_connection._CallbackResult object at 0x7f6ffa0e3948>>, 'arguments': None}
DEBUG      2015-12-20 15:53:19,331 pika.callback                  add                                  164 : Added: {'one_shot': True, 'calls': 1, 'only': None, 'callback': <bound method SelectConnection._on_connection_start of <pika.adapters.select_connection.SelectConnection object at 0x7f6ffd177d30>>, 'arguments': None}
INFO       2015-12-20 15:53:19,333 pika.adapters.base_connection  _create_and_connect_to_socket        212 : Connecting to 127.0.0.1:5672
DEBUG      2015-12-20 15:53:19,334 pika.callback                  process                              220 : Processing 0:Connection.Start
DEBUG      2015-12-20 15:53:19,334 pika.callback                  _use_one_shot_callback               404 : Processing use of oneshot callback
DEBUG      2015-12-20 15:53:19,334 pika.callback                  _use_one_shot_callback               406 : 0 registered uses left
DEBUG      2015-12-20 15:53:19,335 pika.callback                  remove                               269 : Removing callback #0: {'one_shot': True, 'calls': 0, 'only': None, 'callback': <bound method SelectConnection._on_connection_start of <pika.adapters.select_connection.SelectConnection object at 0x7f6ffd177d30>>, 'arguments': None}
DEBUG      2015-12-20 15:53:19,335 pika.callback                  process                              234 : Calling <bound method SelectConnection._on_connection_start of <pika.adapters.select_connection.SelectConnection object at 0x7f6ffd177d30>> for "0:Connection.Start"
DEBUG      2015-12-20 15:53:19,335 pika.callback                  add                                  164 : Added: {'one_shot': True, 'calls': 1, 'only': None, 'callback': <bound method SelectConnection._on_connection_tune of <pika.adapters.select_connection.SelectConnection object at 0x7f6ffd177d30>>, 'arguments': None}
DEBUG      2015-12-20 15:53:19,335 pika.callback                  process                              220 : Processing 0:Connection.Tune
DEBUG      2015-12-20 15:53:19,335 pika.callback                  _use_one_shot_callback               404 : Processing use of oneshot callback
DEBUG      2015-12-20 15:53:19,335 pika.callback                  _use_one_shot_callback               406 : 0 registered uses left
DEBUG      2015-12-20 15:53:19,335 pika.callback                  remove                               269 : Removing callback #0: {'one_shot': True, 'calls': 0, 'only': None, 'callback': <bound method SelectConnection._on_connection_tune of <pika.adapters.select_connection.SelectConnection object at 0x7f6ffd177d30>>, 'arguments': None}
DEBUG      2015-12-20 15:53:19,335 pika.callback                  process                              234 : Calling <bound method SelectConnection._on_connection_tune of <pika.adapters.select_connection.SelectConnection object at 0x7f6ffd177d30>> for "0:Connection.Tune"
DEBUG      2015-12-20 15:53:19,336 pika.connection                _create_heartbeat_checker            1025: Creating a HeartbeatChecker: 580
DEBUG      2015-12-20 15:53:19,336 pika.callback                  add                                  164 : Added: {'one_shot': True, 'calls': 1, 'only': None, 'callback': <bound method SelectConnection._on_connection_open of <pika.adapters.select_connection.SelectConnection object at 0x7f6ffd177d30>>, 'arguments': None}
DEBUG      2015-12-20 15:53:19,336 pika.callback                  process                              220 : Processing 0:Connection.OpenOk
DEBUG      2015-12-20 15:53:19,336 pika.callback                  _use_one_shot_callback               404 : Processing use of oneshot callback
DEBUG      2015-12-20 15:53:19,336 pika.callback                  _use_one_shot_callback               406 : 0 registered uses left
DEBUG      2015-12-20 15:53:19,336 pika.callback                  remove                               269 : Removing callback #0: {'one_shot': True, 'calls': 0, 'only': None, 'callback': <bound method SelectConnection._on_connection_open of <pika.adapters.select_connection.SelectConnection object at 0x7f6ffd177d30>>, 'arguments': None}
DEBUG      2015-12-20 15:53:19,336 pika.callback                  process                              234 : Calling <bound method SelectConnection._on_connection_open of <pika.adapters.select_connection.SelectConnection object at 0x7f6ffd177d30>> for "0:Connection.OpenOk"
DEBUG      2015-12-20 15:53:19,336 pika.callback                  add                                  164 : Added: {'one_shot': True, 'calls': 1, 'only': None, 'callback': <bound method SelectConnection._on_connection_closed of <pika.adapters.select_connection.SelectConnection object at 0x7f6ffd177d30>>, 'arguments': None}
DEBUG      2015-12-20 15:53:19,337 pika.callback                  process                              220 : Processing 0:_on_connection_open
DEBUG      2015-12-20 15:53:19,337 pika.callback                  process                              234 : Calling <bound method _CallbackResult.set_value_once of <pika.adapters.blocking_connection._CallbackResult object at 0x7f6ffb834ec8>> for "0:_on_connection_open"
Stopping and halting node rabbit@hal ...
DEBUG      2015-12-20 15:53:20,703 pika.connection                _create_channel                      1013: Creating channel 1
DEBUG      2015-12-20 15:53:20,704 pika.callback                  add                                  164 : Added: {'one_shot': True, 'calls': 1, 'only': <pika.channel.Channel object at 0x7f6ff9c1fd30>, 'callback': <bound method SelectConnection._on_channel_cleanup of <pika.adapters.select_connection.SelectConnection object at 0x7f6ffd177d30>>, 'arguments': None}
DEBUG      2015-12-20 15:53:20,704 pika.callback                  add                                  164 : Added: {'one_shot': False, 'only': None, 'callback': <bound method Channel._on_getempty of <pika.channel.Channel object at 0x7f6ff9c1fd30>>, 'arguments': None}
DEBUG      2015-12-20 15:53:20,704 pika.callback                  add                                  164 : Added: {'one_shot': False, 'only': None, 'callback': <bound method Channel._on_cancel of <pika.channel.Channel object at 0x7f6ff9c1fd30>>, 'arguments': None}
DEBUG      2015-12-20 15:53:20,704 pika.callback                  add                                  164 : Added: {'one_shot': False, 'only': None, 'callback': <bound method Channel._on_flow of <pika.channel.Channel object at 0x7f6ff9c1fd30>>, 'arguments': None}
DEBUG      2015-12-20 15:53:20,705 pika.callback                  add                                  164 : Added: {'one_shot': True, 'calls': 1, 'only': None, 'callback': <bound method Channel._on_close of <pika.channel.Channel object at 0x7f6ff9c1fd30>>, 'arguments': None}
DEBUG      2015-12-20 15:53:20,705 pika.channel                   _rpc                                 1130: Adding in on_synchronous_complete callback
DEBUG      2015-12-20 15:53:20,705 pika.callback                  add                                  164 : Added: {'one_shot': True, 'calls': 1, 'only': None, 'callback': <bound method Channel._on_synchronous_complete of <pika.channel.Channel object at 0x7f6ff9c1fd30>>, 'arguments': None}
DEBUG      2015-12-20 15:53:20,705 pika.channel                   _rpc                                 1135: Adding passed in callback
DEBUG      2015-12-20 15:53:20,706 pika.callback                  add                                  164 : Added: {'one_shot': True, 'calls': 1, 'only': None, 'callback': <bound method Channel._on_openok of <pika.channel.Channel object at 0x7f6ff9c1fd30>>, 'arguments': None}
ERROR      2015-12-20 15:53:20,706 pika.adapters.base_connection  _handle_error                        344 : Socket Error: 104
DEBUG      2015-12-20 15:53:20,706 pika.heartbeat                 stop                                 108 : Removing timeout for next heartbeat interval
WARNING    2015-12-20 15:53:20,706 pika.adapters.base_connection  _check_state_on_disconnect           180 : Socket closed when connection was open
DEBUG      2015-12-20 15:53:20,706 pika.callback                  add                                  164 : Added: {'one_shot': True, 'calls': 1, 'only': None, 'callback': <bound method SelectConnection._on_connection_start of <pika.adapters.select_connection.SelectConnection object at 0x7f6ffd177d30>>, 'arguments': None}
WARNING    2015-12-20 15:53:20,707 pika.connection                _on_disconnect                       1382: Disconnected from RabbitMQ at localhost:5672 (0): Not specified
DEBUG      2015-12-20 15:53:20,707 pika.callback                  process                              220 : Processing 0:_on_connection_closed
DEBUG      2015-12-20 15:53:20,707 pika.callback                  process                              234 : Calling <bound method _CallbackResult.set_value_once of <pika.adapters.blocking_connection._CallbackResult object at 0x7f6ffa0e3948>> for "0:_on_connection_closed"
Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "/mnt/DATA/greg/work/public/python/pika/pika/adapters/blocking_connection.py", line 698, in channel
    channel_number=channel_number)
  File "/mnt/DATA/greg/work/public/python/pika/pika/connection.py", line 733, in channel
    return self._channels[channel_number]
KeyError: 1
```

gst commented:
with python2.7 : 
```
$ python2.7 -c "import logging ; LOG_FORMAT = ('%(levelname) -10s %(asctime)s %(name) -30s %(funcName) -35s %(lineno) -5d: %(message)s') ; logging.basicConfig(level=logging.DEBUG, format=LOG_FORMAT); import pika; print(pika); c = pika.BlockingConnection(); import subprocess; subprocess.check_call(['sudo', 'rabbitmqctl', 'stop']); c.channel()"
<module 'pika' from 'pika/__init__.pyc'>
DEBUG      2015-12-20 15:54:31,916 pika.adapters.select_connection _get_poller                          137 : Using EPollPoller
DEBUG      2015-12-20 15:54:31,916 pika.callback                  add                                  164 : Added: {'callback': <bound method _CallbackResult.set_value_once of <pika.adapters.blocking_connection._CallbackResult object at 0x7f4cdba50f80>>, 'only': None, 'one_shot': False, 'arguments': None}
DEBUG      2015-12-20 15:54:31,916 pika.callback                  add                                  164 : Added: {'callback': <bound method _CallbackResult.set_value_once of <pika.adapters.blocking_connection._CallbackResult object at 0x7f4cdd677098>>, 'only': None, 'one_shot': False, 'arguments': None}
DEBUG      2015-12-20 15:54:31,916 pika.callback                  add                                  164 : Added: {'callback': <bound method _CallbackResult.set_value_once of <pika.adapters.blocking_connection._CallbackResult object at 0x7f4cdbeac320>>, 'only': None, 'one_shot': False, 'arguments': None}
DEBUG      2015-12-20 15:54:31,917 pika.callback                  add                                  164 : Added: {'callback': <bound method SelectConnection._on_connection_start of <pika.adapters.select_connection.SelectConnection object at 0x7f4cdbeaabd0>>, 'only': None, 'one_shot': True, 'arguments': None, 'calls': 1}
INFO       2015-12-20 15:54:31,917 pika.adapters.base_connection  _create_and_connect_to_socket        212 : Connecting to 127.0.0.1:5672
DEBUG      2015-12-20 15:54:31,918 pika.callback                  process                              220 : Processing 0:Connection.Start
DEBUG      2015-12-20 15:54:31,918 pika.callback                  _use_one_shot_callback               404 : Processing use of oneshot callback
DEBUG      2015-12-20 15:54:31,918 pika.callback                  _use_one_shot_callback               406 : 0 registered uses left
DEBUG      2015-12-20 15:54:31,918 pika.callback                  remove                               269 : Removing callback #0: {'callback': <bound method SelectConnection._on_connection_start of <pika.adapters.select_connection.SelectConnection object at 0x7f4cdbeaabd0>>, 'only': None, 'one_shot': True, 'arguments': None, 'calls': 0}
DEBUG      2015-12-20 15:54:31,919 pika.callback                  process                              234 : Calling <bound method SelectConnection._on_connection_start of <pika.adapters.select_connection.SelectConnection object at 0x7f4cdbeaabd0>> for "0:Connection.Start"
DEBUG      2015-12-20 15:54:31,919 pika.callback                  add                                  164 : Added: {'callback': <bound method SelectConnection._on_connection_tune of <pika.adapters.select_connection.SelectConnection object at 0x7f4cdbeaabd0>>, 'only': None, 'one_shot': True, 'arguments': None, 'calls': 1}
DEBUG      2015-12-20 15:54:31,919 pika.callback                  process                              220 : Processing 0:Connection.Tune
DEBUG      2015-12-20 15:54:31,919 pika.callback                  _use_one_shot_callback               404 : Processing use of oneshot callback
DEBUG      2015-12-20 15:54:31,919 pika.callback                  _use_one_shot_callback               406 : 0 registered uses left
DEBUG      2015-12-20 15:54:31,919 pika.callback                  remove                               269 : Removing callback #0: {'callback': <bound method SelectConnection._on_connection_tune of <pika.adapters.select_connection.SelectConnection object at 0x7f4cdbeaabd0>>, 'only': None, 'one_shot': True, 'arguments': None, 'calls': 0}
DEBUG      2015-12-20 15:54:31,920 pika.callback                  process                              234 : Calling <bound method SelectConnection._on_connection_tune of <pika.adapters.select_connection.SelectConnection object at 0x7f4cdbeaabd0>> for "0:Connection.Tune"
DEBUG      2015-12-20 15:54:31,920 pika.connection                _create_heartbeat_checker            1025: Creating a HeartbeatChecker: 580
DEBUG      2015-12-20 15:54:31,920 pika.callback                  add                                  164 : Added: {'callback': <bound method SelectConnection._on_connection_open of <pika.adapters.select_connection.SelectConnection object at 0x7f4cdbeaabd0>>, 'only': None, 'one_shot': True, 'arguments': None, 'calls': 1}
DEBUG      2015-12-20 15:54:31,920 pika.callback                  process                              220 : Processing 0:Connection.OpenOk
DEBUG      2015-12-20 15:54:31,920 pika.callback                  _use_one_shot_callback               404 : Processing use of oneshot callback
DEBUG      2015-12-20 15:54:31,920 pika.callback                  _use_one_shot_callback               406 : 0 registered uses left
DEBUG      2015-12-20 15:54:31,920 pika.callback                  remove                               269 : Removing callback #0: {'callback': <bound method SelectConnection._on_connection_open of <pika.adapters.select_connection.SelectConnection object at 0x7f4cdbeaabd0>>, 'only': None, 'one_shot': True, 'arguments': None, 'calls': 0}
DEBUG      2015-12-20 15:54:31,920 pika.callback                  process                              234 : Calling <bound method SelectConnection._on_connection_open of <pika.adapters.select_connection.SelectConnection object at 0x7f4cdbeaabd0>> for "0:Connection.OpenOk"
DEBUG      2015-12-20 15:54:31,921 pika.callback                  add                                  164 : Added: {'callback': <bound method SelectConnection._on_connection_closed of <pika.adapters.select_connection.SelectConnection object at 0x7f4cdbeaabd0>>, 'only': None, 'one_shot': True, 'arguments': None, 'calls': 1}
DEBUG      2015-12-20 15:54:31,921 pika.callback                  process                              220 : Processing 0:_on_connection_open
DEBUG      2015-12-20 15:54:31,921 pika.callback                  process                              234 : Calling <bound method _CallbackResult.set_value_once of <pika.adapters.blocking_connection._CallbackResult object at 0x7f4cdd677098>> for "0:_on_connection_open"
Stopping and halting node rabbit@hal ...
DEBUG      2015-12-20 15:54:33,282 pika.connection                _create_channel                      1013: Creating channel 1
DEBUG      2015-12-20 15:54:33,283 pika.callback                  add                                  164 : Added: {'callback': <bound method SelectConnection._on_channel_cleanup of <pika.adapters.select_connection.SelectConnection object at 0x7f4cdbeaabd0>>, 'only': <pika.channel.Channel object at 0x7f4cdd742750>, 'one_shot': True, 'arguments': None, 'calls': 1}
DEBUG      2015-12-20 15:54:33,283 pika.callback                  add                                  164 : Added: {'callback': <bound method Channel._on_getempty of <pika.channel.Channel object at 0x7f4cdd742750>>, 'only': None, 'one_shot': False, 'arguments': None}
DEBUG      2015-12-20 15:54:33,284 pika.callback                  add                                  164 : Added: {'callback': <bound method Channel._on_cancel of <pika.channel.Channel object at 0x7f4cdd742750>>, 'only': None, 'one_shot': False, 'arguments': None}
DEBUG      2015-12-20 15:54:33,284 pika.callback                  add                                  164 : Added: {'callback': <bound method Channel._on_flow of <pika.channel.Channel object at 0x7f4cdd742750>>, 'only': None, 'one_shot': False, 'arguments': None}
DEBUG      2015-12-20 15:54:33,284 pika.callback                  add                                  164 : Added: {'callback': <bound method Channel._on_close of <pika.channel.Channel object at 0x7f4cdd742750>>, 'only': None, 'one_shot': True, 'arguments': None, 'calls': 1}
DEBUG      2015-12-20 15:54:33,285 pika.channel                   _rpc                                 1130: Adding in on_synchronous_complete callback
DEBUG      2015-12-20 15:54:33,285 pika.callback                  add                                  164 : Added: {'callback': <bound method Channel._on_synchronous_complete of <pika.channel.Channel object at 0x7f4cdd742750>>, 'only': None, 'one_shot': True, 'arguments': None, 'calls': 1}
DEBUG      2015-12-20 15:54:33,285 pika.channel                   _rpc                                 1135: Adding passed in callback
DEBUG      2015-12-20 15:54:33,285 pika.callback                  add                                  164 : Added: {'callback': <bound method Channel._on_openok of <pika.channel.Channel object at 0x7f4cdd742750>>, 'only': None, 'one_shot': True, 'arguments': None, 'calls': 1}
ERROR      2015-12-20 15:54:33,285 pika.adapters.base_connection  _handle_error                        344 : Socket Error: 104
DEBUG      2015-12-20 15:54:33,285 pika.heartbeat                 stop                                 108 : Removing timeout for next heartbeat interval
WARNING    2015-12-20 15:54:33,286 pika.adapters.base_connection  _check_state_on_disconnect           180 : Socket closed when connection was open
DEBUG      2015-12-20 15:54:33,286 pika.callback                  add                                  164 : Added: {'callback': <bound method SelectConnection._on_connection_start of <pika.adapters.select_connection.SelectConnection object at 0x7f4cdbeaabd0>>, 'only': None, 'one_shot': True, 'arguments': None, 'calls': 1}
WARNING    2015-12-20 15:54:33,286 pika.connection                _on_disconnect                       1382: Disconnected from RabbitMQ at localhost:5672 (0): Not specified
DEBUG      2015-12-20 15:54:33,286 pika.callback                  process                              220 : Processing 0:_on_connection_closed
DEBUG      2015-12-20 15:54:33,286 pika.callback                  process                              234 : Calling <bound method _CallbackResult.set_value_once of <pika.adapters.blocking_connection._CallbackResult object at 0x7f4cdbeac320>> for "0:_on_connection_closed"
Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "pika/adapters/blocking_connection.py", line 698, in channel
    channel_number=channel_number)
  File "pika/connection.py", line 733, in channel
    return self._channels[channel_number]
KeyError: 1
```

vitaly-krugl commented:
@gst - thx!

vitaly-krugl commented:
@gst, I am able to reproduce this on Fedora with Linux 3.17.4 and python 3.4:

![screen shot 2015-12-20 at 1 30 30 pm](https://cloud.githubusercontent.com/assets/1441066/11920243/6763a4fc-a71e-11e5-9daa-1d44bdf03dfc.png)

This was most likely introduced by this change: https://github.com/pika/pika/pull/578/files#diff-1a5308d8b684fdafac1f89e376e8933dR267; prior to this change, the writing to the socket and the corresponding detection of "errno 104 connection reset by peer" would have occurred later on in the context of the io-loop (after `Channel.open()` returned) 

CC @wjps

vitaly-krugl commented:
I can also reproduce the KeyError exception with python 2.7.8 on my Fedora VirtualBox instance.

I wish I understood why `TestNoAccessToFileDescriptorAfterConnectionClosed` ( https://github.com/pika/pika/blob/94f86187bf8cbd6cdc18a52f17c564fbb2c78169/tests/acceptance/blocking_adapter_test.py#L196-L218) doesn't catch this problem in pika's Travis builds.

vitaly-krugl commented on pika/connection.py:
```diff
@@ -704,11 +704,11 @@ def channel(self, on_open_callback, channel_number=None):
             channel_number = self._next_channel_number()
-        self._channels[channel_number] = self._create_channel(channel_number,
-                                                              on_open_callback)
+        channel = self._channels[channel_number] = self._create_channel(channel_number,
```
Need to stay within 80 characters per line.

vitaly-krugl commented:
@gst: besides the 80-char/line limit feedback, your fix in `Connection.channel()` looks reasonable. The only other thing that I recommend is to add a comment before `channel = ` that explains why the code is being careful with the index-based access. It's not obvious why this is being done, and some other maintainer could easily "optimize" the access back to the way it was before your change.

vitaly-krugl commented:
I do wish that we had a test that could reproduce the issue in pika test suite. I am also concerned that there might be additional fallout from https://github.com/pika/pika/pull/578/files#diff-1a5308d8b684fdafac1f89e376e8933dR267. A code review may be necessary to identify additional flaws like the one that you fixed.

gst commented:
> I do wish that we had a test that could reproduce the issue in pika test suite.

there I added one. I tried to use mock to trigger the send() to raise socket.error but couldn't succeed.. after 2 minutes I got this self-made solution.. don't know if it's ok to you.
I also added comment within the channel method to let the reader know..

vitaly-krugl commented on tests/unit/blocking_connection_tests.py:
```diff
@@ -5,6 +5,10 @@
 
+import errno
+
+from pika.exceptions import ConnectionClosed
```
A commonly-used practice is to group imports as follows for readability:

1. Imports of builtins
2. Imports of 3rd party modules
3. Imports of our own modules

So, this new pika.exceptions import belongs with the other pika imports below

Also, all other code in this test suite accesses pika exceptions relative to their namespace, so to follow that convention, your code would access it as `pika.exceptions.ConnectionClosed`. However, we still need an explicit import `import pika.exceptions` after `from pika.adapters import ...`. The original implementation forgot to add that import and got by implicitly.

vitaly-krugl commented on tests/unit/blocking_connection_tests.py:
```diff
@@ -5,6 +5,10 @@
 """
 import socket
 
+import errno
```
It's easier on the eyes (for some developers, at least :) ) if the imports within each group are sorted, so in this case, you might have:

```
import errno
import socket
```

vitaly-krugl commented on tests/unit/blocking_connection_tests.py:
```diff
@@ -204,3 +208,33 @@ def test_sleep(self, select_connection_class_mock):
+            # initiate a connection:
+            conn = blocking_connection.BlockingConnection()
+
+            # then wraps it within a proxy so to be able
```
This comment seems incomplete: to be able ... to do what?

gst replied:
effectively ;) the next comment (not far away) resumes this one..

vitaly-krugl commented on tests/unit/blocking_connection_tests.py:
```diff
@@ -204,3 +208,33 @@ def test_sleep(self, select_connection_class_mock):
+
+        for error in (errno.EPIPE, errno.ECONNABORTED, errno.ECONNREFUSED):
+            # initiate a connection:
+            conn = blocking_connection.BlockingConnection()
```
This relies on actually being able to connect to a live broker. Unit tests are not allowed to do that. All external networking needs to be mocked out as in the other unit tests. Unit tests are expected to work even when the external servers are not running.

gst replied:
yeah, that was the easier solution I found actually.
I agree with most of your comments/remarks but otherwise I would have to handle a full socket(.socket) mock which was harder to me. I just found a lib to do that, giving it a look right now.. (https://github.com/mocketize/python-mocket).. would it makes the job so here ?

vitaly-krugl replied:
We can probably handle mock testing at each layer without introducing a new dependency. It doesn't seem to be documented or even very active. I don't  I don't know whether Gavin (@gmr) would agree to this dependency. You might want to check with him before investing the time.

gst replied:
I agree it's not good ;) à simple tcpserver would do the trick..

gst commented:
by the way my test function was even not testing the actual exception that I receive in real situation.. the actual error received (in `SelectConnection._handle_write()`) is ECONNRESET (104) .. which makes this code path, in `_handle_error()`, to actually be covered:
```python
        else:
            # Haven't run into this one yet, log it.
            LOGGER.error("Socket Error: %s", error_code)
```
I'll review all that..  though now I'm unsure the python-mocket lib would make my life easier here.. have to give it a try..

vitaly-krugl commented:
Regarding `# Haven't run into this one yet, log it`, if it needs to be tested, it should be tested at its own layer in `base_connection_tests.py`

gst commented:
I have a test case always reproducing the problem, it's not using a real server but a simple threaded tcpserver (launched from within the test) which upon connection from the client I then write & read the expected packets to the client (amqp hanshake and so on), syncing between the client and server with thread events (so to be sure to have the right sequence of actions ongoing).. 

if you will I clean & submit it  ?

I do think the same could be used to reproduce quite a lot of others problems of the same kind..

gst commented:
> I have a test case always reproducing the problem
I thought it was always but apparently not.. it depends on I don't know (sometimes the write succeed (after the connection was closed :s) and I get read to return 0 data, sometimes I get write fails with ECONNRESET, etc..).. by also using a proxy object over the client socket and forcing the right exception for the send() implied in channel() creation now it's so really always reproducing.

vitaly-krugl commented:
@gst, I also discovered that it often takes more than one write to a remotely-closed socket to get the error. That's why the test https://github.com/pika/pika/blob/94f86187bf8cbd6cdc18a52f17c564fbb2c78169/tests/acceptance/blocking_adapter_test.py#L196 was not reproducing the problem.

gst commented:
there it is.

vitaly-krugl commented on tests/unit/blocking_connection_tests.py:
```diff
@@ -204,3 +216,149 @@ def test_sleep(self, select_connection_class_mock):
             connection.sleep(0.00001)
 
 
+HAS_SO_LINGER = hasattr(socket, 'SO_LINGER')
```
Which platforms don't have `SO_LINGER`?

gst replied:
non POSIX ones I guess ?

gst replied:
> but if you think it's not necessary to test for its presence then I can unconditionnaly use it, no problem..

that would remove the need (though, I do think it's better to use it, even always) of the _Proxy() "mock" needed to be sure to trigger **exactly** the expected error/exception..

vitaly-krugl commented:
@gst, as I suspected, https://github.com/pika/pika/pull/578/files#diff-1a5308d8b684fdafac1f89e376e8933dR267 messed up more than just the `Connection.channel()` case that this PR is trying to address. For example, `Connection._on_connection_tune` gets into trouble if connection reset is discovered while executing in `Connection._send_connection_tune_ok`, when `BaseConnection._flush_outbound` is calling `BaseConnection._handle_write`. In that case, before `Connection._on_connection_tune` calls `Connection._send_connection_open`, the connection is already in the closed state and the `ON_ERROR_CALLBACK` has already been sent. However, `_on_connection_tune` still goes on to call `_send_connection_open`, which is not happy about being called in the closed state and trips up on a non-specific `ConnectionClosed` exception inside `Connection._send_frame`.

All this is happening because pika's code was not originally designed to be writing to (or reading from) the socket from any methods, except from ioloop callbacks. Unfortunately, the PR that contains the change https://github.com/pika/pika/pull/578/files#diff-1a5308d8b684fdafac1f89e376e8933dR267, didn't take that into account, causing all sorts of havoc in `Connection.channel()`, `Connection._on_connection_tune`, etc.

There are certainly likely to be other places in pika that won't behave properly under the circumstances as the result of that change. It's a slippery slope.

To make the long story short, I think that at this time, it's more practical to do the following:

1. Undo all the changes made so far in this PR. It's a same after all this effort, but there is no other practical choice.
2. Remove the self._handle_write() call from `BaseConnection._flush_outbound` and restore the docstring to what it was before that change. Also, add a comment in `BaseConnection._flush_outbound` for future maintainers, describing why its presently not possible to call `self._handle_write()` in that context - essentially, something like this "pika code was not designed to be writing to (or reading from) the socket from any methods, except from ioloop handler callbacks".
 * https://github.com/pika/pika/pull/578/files#diff-1a5308d8b684fdafac1f89e376e8933dR267

vitaly-krugl commented:
@gst, I will be offline for a few days. Please see my latest recommendation in my previous comment. Best regards and many thanks for your effort.

vitaly-krugl commented:
@gst, per my prior comment, I submitted a new PR #679 with a more general fix that resolves #659 as well as other similar bugs and also updates the `blocking_adapter_test.py` acceptance test suite to verify the fix in the `Connection.channel` case. You work here helped me figure out how to reproduce the issue with the help of `SO_LINGER` in the ForwardServer test utility.

Would you mind reviewing the new PR #679?

gst commented:
@vitaly-krugl  :  OK I, more or less, get what you've discovered / are explaining. 
Gonna close this PR so, no problem.

and giving the #679 a review.. (but it will be more a diagonal one, the PR not being a little one ;)

vitaly-krugl commented:
thx
